### PR TITLE
MRG: RSA example

### PR DESCRIPTION
**Pull request by agramfort**

### CODE REVIEW

**Comment by agramfort** [comment]
here is what it gives me:

screen shot 2017-01-20 at 18 35 39

**Comment by agramfort** [comment]
@kingjr please have a look

File: examples/decoding/decoding_rsa.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+[2] Laakso, A. & Cottrell, G.. "Content and cluster analysis:
+ assessing representational similarity in neural systems." Philosophical
+ psychology 13.1 (2000): 47-76.
+
+[3] Kriegeskorte, N., Marieke, M., & Bandettini.  P. "Representational
+ similarity analysis-connecting the branches of systems neuroscience."
+ Frontiers in systems neuroscience 2 (2008): 4.
+
+[4] Cichy, R. M., Pantazis, D., & Oliva, A. "Resolving human object recognition
+in space and time. Nature neuroscience" (2014): 17(3), 455-462
```
Comment (Eric89GXL) [inline examples/decoding/decoding_rsa.py]: these refs are formatted incorrectly, need `[1]_` above and `.. [1]` here

File: examples/decoding/decoding_rsa.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+max_trigger = 24
+conds = conds[:max_trigger]  # take only the first 24 rows
+
+##############################################################################
+# Define stimulus - trigger mapping
+conditions = ['/'.join(map(str,
+                           list(c[:2]) + [('not-' if i == 0 else '') +
+                                          conds.columns[k]
+                                          for k, i in enumerate(c[2:], 2)]))
+              for c in conds.values]
```
Comment (Eric89GXL) [inline examples/decoding/decoding_rsa.py]: This would be a lot more readable in loop form

File: examples/decoding/decoding_rsa.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+print(conditions[:10])
+
+##############################################################################
+# Let's make the event_id dictionary
+event_id = dict(zip(conditions, map(int, conds.trigger + 1)))
```
Comment (Eric89GXL) [inline examples/decoding/decoding_rsa.py]: why does this need to be mapped? is `np.int` not supported by `event_id`?

File: examples/decoding/decoding_rsa.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+events = mne.find_events(raw, min_duration=.002, verbose=True)
+print(events[:10])
+
+events = events[events[:, 2] <= max_trigger]
+mne.viz.plot_events(events, sfreq=raw.info['sfreq'])
+
+##############################################################################
+# Epoch data
+picks = mne.pick_types(raw.info, meg=True)
```
Comment (Eric89GXL) [inline examples/decoding/decoding_rsa.py]: why do picking? Is there EEG data that is being excluded or something?

File: examples/decoding/decoding_rsa.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+# all pairwise binary classification task to make the RDM.
+# Also we use here the ROC-AUC as performance metric while the
+# paper uses accuracy. Finally here for the sake of time we use
+# RSA on a window of data while Cichy et al. did it for all time
+# instants separately.
+
+# Classify using the average signal in the window 50ms to 300ms
+clf = make_pipeline(StandardScaler(),
+                    LogisticRegression(C=1, solver='lbfgs'))
+X = epochs.copy().crop(0.05, 0.3).get_data().mean(axis=2)
```
Comment (Eric89GXL) [inline examples/decoding/decoding_rsa.py]: add comment like `# use time-average over the 50-300 ms interval`

Comment (Eric89GXL) [inline examples/decoding/decoding_rsa.py]: (for speed?)

File: examples/decoding/decoding_rsa.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
+"""
+====================================
+Representational Similarity Analysis
+====================================
+
+Representational Similarity Analysis is used to perform summary statistics
+on supervised classifications where the number of classes is relatively high.
+It consists in characterizing the structure of the confusion matrix to infer
+the similarity between brain responses and serve as a proxy for characterizing
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: serves

File: examples/decoding/decoding_rsa.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+the space of mental representations [1-3].
+
+In this example, we perform RSA on responses to 92 object images. Subjects were
+presented images of human, animal and inanimate objects [4].
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: presented with

File: examples/decoding/decoding_rsa.py (hunk 2 above)
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: +1, weird alignment

File: examples/decoding/decoding_rsa.py (hunk 7 above)
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: 24 distinct images of faces and body parts

File: examples/decoding/decoding_rsa.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+n_runs = 4  # 4 for full data (use less to speed up computations)
+fname = op.join(data_path, 'sample_subject_%i_tsss_mc.fif')
+raws = [read_raw_fif(fname % block) for block in range(n_runs)]
+raw = concatenate_raws(raws)
+
+events = mne.find_events(raw, min_duration=.002, verbose=True)
+print(events[:10])
+
+events = events[events[:, 2] <= max_trigger]
+mne.viz.plot_events(events, sfreq=raw.info['sfreq'])
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: necessary?

File: examples/decoding/decoding_rsa.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+##############################################################################
+# Read MEG data
+n_runs = 4  # 4 for full data (use less to speed up computations)
+fname = op.join(data_path, 'sample_subject_%i_tsss_mc.fif')
+raws = [read_raw_fif(fname % block) for block in range(n_runs)]
+raw = concatenate_raws(raws)
+
+events = mne.find_events(raw, min_duration=.002, verbose=True)
+print(events[:10])
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: necessary?

File: examples/decoding/decoding_rsa.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+event_id['0/human bodypart/human/not-face/animal/natural']
+
+##############################################################################
+# Read MEG data
+n_runs = 4  # 4 for full data (use less to speed up computations)
+fname = op.join(data_path, 'sample_subject_%i_tsss_mc.fif')
+raws = [read_raw_fif(fname % block) for block in range(n_runs)]
+raw = concatenate_raws(raws)
+
+events = mne.find_events(raw, min_duration=.002, verbose=True)
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: remove verbose?

File: examples/decoding/decoding_rsa.py (hunk 4 above)
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: Since the classifier deals with a numpy array, this ensures that we don't include trigger channels

File: examples/decoding/decoding_rsa.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+# classifier (Multinomial Logistic Regression) while the paper uses
+# all pairwise binary classification task to make the RDM.
+# Also we use here the ROC-AUC as performance metric while the
+# paper uses accuracy. Finally here for the sake of time we use
+# RSA on a window of data while Cichy et al. did it for all time
+# instants separately.
+
+# Classify using the average signal in the window 50ms to 300ms
+clf = make_pipeline(StandardScaler(),
+                    LogisticRegression(C=1, solver='lbfgs'))
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: does the solver make a big diff?

File: examples/decoding/decoding_rsa.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+X = epochs.copy().crop(0.05, 0.3).get_data().mean(axis=2)
+y = epochs.events[:, 2]
+
+n_folds = 5
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: remove and directly specify n_splits in cv

File: examples/decoding/decoding_rsa.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+classes = set(y)
+cv = StratifiedKFold(n_folds, random_state=0, shuffle=True)
+
+# Compute confusion matrix for each cross-validation fold
+y_pred = np.zeros((len(y), len(classes)))
+for train, test in cv.split(X, y):
+    # Fit
+    clf.fit(X[train], y[train])
+    # Predict
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: Probabilistic prediction

File: examples/decoding/decoding_rsa.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    y_pred[test] = clf.predict_proba(X[test])
+
+print(y_pred[:2])
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: remove?

File: examples/decoding/decoding_rsa.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+
+##############################################################################
+# Compute confusion matrix using ROC-AUC
+cm = np.zeros((len(classes), len(classes)))
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: maybe cm => confusion

File: examples/decoding/decoding_rsa.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+for ii, train_class in enumerate(classes):
+    for jj in range(ii, len(classes)):
+        cm[ii, jj] = roc_auc_score(y == train_class, y_pred[:, jj])
+        cm[jj, ii] = cm[ii, jj]
+
+# Format class names for centered plotting
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: oh dear this is complicated for an example. How about a manual:
`labels = [[''] * 6 + ['face'] + [''] * 12 + ['body'] + [''] * 6]`
ax.set_ytickslabel(labels)

File: examples/decoding/decoding_rsa.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+fig, ax = plt.subplots(1)
+im = ax.matshow(cm, cmap='RdBu_r', clim=[0.3, 0.7])
+ax.set_yticks(range(len(classes)))
+ax.set_yticklabels(sparse_names)
+ax.set_xticks(range(len(classes)))
+ax.set_xticklabels(sparse_names, rotation=40, ha='left')
+for ii, (name, next_name) in enumerate(zip(names, names[1:])):
+    if name != next_name:
+        ax.axhline(ii + 0.5, color='k')
+        ax.axvline(ii + 0.5, color='k')
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: same here, I would simplify to a manual loop, no need to be too smart with that number of categories. (my bad)

File: examples/decoding/decoding_rsa.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+##############################################################################
+# Confusion matrix related to mental representations have been historically
+# summarized with dimensionality reduction using multi-dimensional scaling [1].
+# See how the face samples cluster together.
+fig, ax = plt.subplots(1)
+mds = MDS(2, random_state=0, dissimilarity='precomputed')
+chance = 0.5
+summary = mds.fit_transform(chance - cm)
+cmap = plt.get_cmap('rainbow')
+colors = cmap(np.linspace(0., 1., len(set(names))))
```
Comment (kingjr) [inline examples/decoding/decoding_rsa.py]: not necessary anymore since there are two colors?

File: examples/decoding/decoding_rsa.py (hunk 3 above)
Comment (agramfort) [inline examples/decoding/decoding_rsa.py]: yes indeed. it's otherwise an np.int64 and isinstance(..., int) is False in this case.

File: examples/decoding/decoding_rsa.py (hunk 5 above)
Comment (agramfort) [inline examples/decoding/decoding_rsa.py]: more performance than speed. It narrows the data to the time interval that has the most signal.

File: examples/decoding/decoding_rsa.py (hunk 8 above)
Comment (agramfort) [inline examples/decoding/decoding_rsa.py]: I see yes it helps understanding the paradigm

File: examples/decoding/decoding_rsa.py (hunk 11 above)
Comment (agramfort) [inline examples/decoding/decoding_rsa.py]: it's the one that supports full multinomial regressoin model

**Comment by agramfort** [comment]
ok i've done what I wanted to complete.

**Comment by agramfort** [comment]
CIs are happy but for some reason my link to the RSA example  from this page does not work:

https://3572-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/manual/datasets_index.html#visual-92-object-categories

I am looking into it. Don't merge yet

**Comment by Eric89GXL** [comment]
Needs rebase now

**Comment by Eric89GXL** [comment]
+1 for merge, ready from your end? If so I can try using the web interface to resolve conflicts

**Comment by agramfort** [comment]
yes good to go for me. See:

https://3573-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_examples/decoding/decoding_rsa.html#sphx-glr-auto-examples-decoding-decoding-rsa-py

**Comment by Eric89GXL** [comment]
Thanks @agramfort

**Comment by agramfort** [comment]
thx

good team effort ;)
